// piral-svelte/lib/converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createConverter = void 0;
const extension_1 = require("./extension");
function createConverter(config = {}) {
    const { selector = 'svelte-extension' } = config;
    const Extension = (0, extension_1.createExtension)(selector);
    const convert = (Component, captured) => ({
        mount(parent, data, ctx, locals) {
            locals.instance = new Component({
                target: parent,
                props: Object.assign(Object.assign(Object.assign({}, captured), ctx), data),
            });
        },
        update(el, data, ctx, locals) {
            Object.keys(data).forEach((key) => {
                locals.instance[key] = data[key];
            });
        },
        unmount(el, locals) {
            locals.instance.$destroy();
            locals.instance = undefined;
            el.innerHTML = '';
        },
    });
    convert.Extension = Extension;
    return convert;
}
exports.createConverter = createConverter;
